diff -u -r ./org/docx4j/wml/Body.java ../patched/org/docx4j/wml/Body.java
--- ./org/docx4j/wml/Body.java	2009-10-10 09:56:01.000000000 +1100
+++ ../patched/org/docx4j/wml/Body.java	2009-10-10 01:22:54.000000000 +1100
@@ -29,6 +29,7 @@
 import javax.xml.bind.annotation.XmlAccessorType;
 import javax.xml.bind.annotation.XmlElementRef;
 import javax.xml.bind.annotation.XmlElementRefs;
+import javax.xml.bind.annotation.XmlRootElement;
 import javax.xml.bind.annotation.XmlTransient;
 import javax.xml.bind.annotation.XmlType;
 import org.docx4j.math.CTOMath;
diff -u -r ./org/docx4j/wml/CTBookmark.java ../patched/org/docx4j/wml/CTBookmark.java
--- ./org/docx4j/wml/CTBookmark.java	2009-10-10 09:56:01.000000000 +1100
+++ ../patched/org/docx4j/wml/CTBookmark.java	2009-07-12 14:55:46.000000000 +1000
@@ -25,6 +25,7 @@
 import javax.xml.bind.annotation.XmlAccessType;
 import javax.xml.bind.annotation.XmlAccessorType;
 import javax.xml.bind.annotation.XmlAttribute;
+import javax.xml.bind.annotation.XmlRootElement;
 import javax.xml.bind.annotation.XmlTransient;
 import javax.xml.bind.annotation.XmlType;
 import org.jvnet.jaxb2_commons.ppp.Child;
@@ -49,6 +50,7 @@
  */
 @XmlAccessorType(XmlAccessType.FIELD)
 @XmlType(name = "CT_Bookmark")
+@XmlRootElement(name = "bookmarkStart")
 public class CTBookmark
     extends CTBookmarkRange
     implements Child
diff -u -r ./org/docx4j/wml/CTMarkupRange.java ../patched/org/docx4j/wml/CTMarkupRange.java
--- ./org/docx4j/wml/CTMarkupRange.java	2009-10-10 09:56:01.000000000 +1100
+++ ../patched/org/docx4j/wml/CTMarkupRange.java	2009-07-12 14:55:46.000000000 +1000
@@ -25,6 +25,7 @@
 import javax.xml.bind.annotation.XmlAccessType;
 import javax.xml.bind.annotation.XmlAccessorType;
 import javax.xml.bind.annotation.XmlAttribute;
+import javax.xml.bind.annotation.XmlRootElement;
 import javax.xml.bind.annotation.XmlTransient;
 import javax.xml.bind.annotation.XmlType;
 import org.jvnet.jaxb2_commons.ppp.Child;
@@ -49,6 +50,8 @@
  */
 @XmlAccessorType(XmlAccessType.FIELD)
 @XmlType(name = "CT_MarkupRange")
+// We don't have @XmlRootElement here, because in addition to bookmarkEnd, there are elements
+// which become this.
 public class CTMarkupRange
     extends CTMarkup
     implements Child
diff -u -r ./org/docx4j/wml/Drawing.java ../patched/org/docx4j/wml/Drawing.java
--- ./org/docx4j/wml/Drawing.java	2009-10-10 09:56:02.000000000 +1100
+++ ../patched/org/docx4j/wml/Drawing.java	2009-10-10 01:22:54.000000000 +1100
@@ -28,6 +28,7 @@
 import javax.xml.bind.annotation.XmlAccessorType;
 import javax.xml.bind.annotation.XmlElement;
 import javax.xml.bind.annotation.XmlElements;
+import javax.xml.bind.annotation.XmlRootElement;
 import javax.xml.bind.annotation.XmlTransient;
 import javax.xml.bind.annotation.XmlType;
 import org.docx4j.dml.Anchor;
@@ -59,13 +60,14 @@
 @XmlType(name = "CT_Drawing", propOrder = {
     "anchorOrInline"
 })
+@XmlRootElement(name ="drawing")
 public class Drawing
     implements Child
 {
 
     @XmlElements({
-        @XmlElement(name = "inline", namespace = "http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing", type = Inline.class),
-        @XmlElement(name = "anchor", namespace = "http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing", type = Anchor.class)
+        @XmlElement(name = "anchor", namespace = "http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing", type = Anchor.class),
+        @XmlElement(name = "inline", namespace = "http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing", type = Inline.class)
     })
     protected List<Object> anchorOrInline;
     @XmlTransient
diff -u -r ./org/docx4j/wml/Id.java ../patched/org/docx4j/wml/Id.java
--- ./org/docx4j/wml/Id.java	2009-10-10 09:56:02.000000000 +1100
+++ ../patched/org/docx4j/wml/Id.java	2008-05-05 17:10:49.000000000 +1000
@@ -115,4 +115,20 @@
         setParent(parent);
     }
 
+    public boolean equals(Object obj) {
+    	if (obj instanceof Id) {
+	    		return val.equals( ((Id)obj).getVal() ); 
+	    	} else {
+	    		return false;
+	    	}
+	    }
+	   
+	    public int hashCode() {
+	    	
+	    	// Natural and good enough...
+	    	return val.intValue();	    	
+	    }
+        
+    
+    
 }
diff -u -r ./org/docx4j/wml/Jc.java ../patched/org/docx4j/wml/Jc.java
--- ./org/docx4j/wml/Jc.java	2009-10-10 09:56:02.000000000 +1100
+++ ../patched/org/docx4j/wml/Jc.java	2009-04-02 23:16:12.000000000 +1100
@@ -25,6 +25,7 @@
 import javax.xml.bind.annotation.XmlAccessType;
 import javax.xml.bind.annotation.XmlAccessorType;
 import javax.xml.bind.annotation.XmlAttribute;
+import javax.xml.bind.annotation.XmlRootElement;
 import javax.xml.bind.annotation.XmlTransient;
 import javax.xml.bind.annotation.XmlType;
 import org.jvnet.jaxb2_commons.ppp.Child;
@@ -49,6 +50,7 @@
  */
 @XmlAccessorType(XmlAccessType.FIELD)
 @XmlType(name = "CT_Jc")
+@XmlRootElement(name = "jc")        
 public class Jc implements Child
 {
 
diff -u -r ./org/docx4j/wml/P.java ../patched/org/docx4j/wml/P.java
--- ./org/docx4j/wml/P.java	2009-10-10 09:56:02.000000000 +1100
+++ ../patched/org/docx4j/wml/P.java	2009-10-10 01:22:54.000000000 +1100
@@ -35,6 +35,7 @@
 import javax.xml.bind.annotation.XmlType;
 import org.docx4j.math.CTOMath;
 import org.docx4j.math.CTOMathPara;
+import org.apache.log4j.Logger;
 import org.jvnet.jaxb2_commons.ppp.Child;
 
 
@@ -673,4 +675,56 @@
 
     }
 
+    // Not generated
+    /** Get the text content of all runs in the P.  WARNING: this only gets ./w:r/w:t.
+     *  That is incomplete.  Try org.docx4j.TextUtils.extractText instead.  */
+    @Deprecated
+    public String toString() {
+
+    	StringBuilder result = new StringBuilder();
+    	
+    	List<Object> children = getParagraphContent();
+    	
+//    	System.out.println("p.toString");
+    	
+		for (Object o : children ) {					
+//			System.out.println("  " + o.getClass().getName() );
+			if ( o instanceof org.docx4j.wml.R) {
+//		    	System.out.println("Hit R");
+				org.docx4j.wml.R  run = (org.docx4j.wml.R)o;
+		    	List runContent = run.getRunContent();
+				for (Object o2 : runContent ) {					
+					if ( o2 instanceof javax.xml.bind.JAXBElement) {
+						// TODO - unmarshall directly to Text.
+						if ( ((JAXBElement)o2).getDeclaredType().getName().equals("org.docx4j.wml.Text") ) {
+//					    	System.out.println("Found Text");
+							org.docx4j.wml.Text t = (org.docx4j.wml.Text)((JAXBElement)o2).getValue();
+							result.append( t.getValue() );					
+						}
+					} else {
+//				    	System.out.println(o2.getClass().getName());						
+					}
+				}
+			} 
+		}
+		return result.toString();
+    	
+    }
+
+    public void replaceElement(Object current, List insertions) {
+
+    	int index = paragraphContent.indexOf(current);    	
+    	if (index > -1 ) {    		
+    		paragraphContent.addAll(index+1, insertions);  
+    		Object removed = paragraphContent.remove(index);
+    		// sanity check
+    		if (!current.equals(removed)) {
+    			log.error("removed wrong object?");
+    		}    		
+    	} else {
+    		// Not found
+    		log.error("Couldn't find replacement target.");
+    	}
+    }    
+    
 }
diff -u -r ./org/docx4j/wml/PPrBase.java ../patched/org/docx4j/wml/PPrBase.java
--- ./org/docx4j/wml/PPrBase.java	2009-10-10 09:56:02.000000000 +1100
+++ ../patched/org/docx4j/wml/PPrBase.java	2009-04-02 12:24:15.000000000 +1100
@@ -26,6 +26,7 @@
 import javax.xml.bind.annotation.XmlAccessType;
 import javax.xml.bind.annotation.XmlAccessorType;
 import javax.xml.bind.annotation.XmlAttribute;
+import javax.xml.bind.annotation.XmlRootElement;
 import javax.xml.bind.annotation.XmlTransient;
 import javax.xml.bind.annotation.XmlType;
 import org.jvnet.jaxb2_commons.ppp.Child;
@@ -1225,6 +1226,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "ind")        
     public static class Ind
         implements Child
     {
@@ -1524,6 +1526,7 @@
         "numberingChange",
         "ins"
     })
+    @XmlRootElement(name = "numPr")    
     public static class NumPr
         implements Child
     {
@@ -1856,6 +1859,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "outlineLvl")        
     public static class OutlineLvl
         implements Child
     {
@@ -1951,6 +1955,7 @@
         "between",
         "bar"
     })
+    @XmlRootElement(name = "pBdr")        
     public static class PBdr
         implements Child
     {
@@ -2244,6 +2249,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "spacing")        
     public static class Spacing
         implements Child
     {
@@ -2325,7 +2331,7 @@
          */
         public boolean isBeforeAutospacing() {
             if (beforeAutospacing == null) {
-                return true;
+                return false;
             } else {
                 return beforeAutospacing;
             }
@@ -2401,7 +2407,7 @@
          */
         public boolean isAfterAutospacing() {
             if (afterAutospacing == null) {
-                return true;
+                return false;
             } else {
                 return afterAutospacing;
             }
diff -u -r ./org/docx4j/wml/PPr.java ../patched/org/docx4j/wml/PPr.java
--- ./org/docx4j/wml/PPr.java	2009-10-10 09:56:02.000000000 +1100
+++ ../patched/org/docx4j/wml/PPr.java	2008-05-01 17:48:11.000000000 +1000
@@ -18,12 +18,12 @@
 
  */
 
-
 package org.docx4j.wml;
 
 import javax.xml.bind.Unmarshaller;
 import javax.xml.bind.annotation.XmlAccessType;
 import javax.xml.bind.annotation.XmlAccessorType;
+import javax.xml.bind.annotation.XmlRootElement;
 import javax.xml.bind.annotation.XmlTransient;
 import javax.xml.bind.annotation.XmlType;
 import org.jvnet.jaxb2_commons.ppp.Child;
@@ -56,6 +56,7 @@
     "sectPr",
     "pPrChange"
 })
+@XmlRootElement(name = "pPr")
 public class PPr
     extends PPrBase
     implements Child
diff -u -r ./org/docx4j/wml/R.java ../patched/org/docx4j/wml/R.java
--- ./org/docx4j/wml/R.java	2009-10-10 09:56:02.000000000 +1100
+++ ../patched/org/docx4j/wml/R.java	2009-10-10 01:22:54.000000000 +1100
@@ -315,6 +315,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "annotationRef")
     public static class AnnotationRef
         implements Child
     {
@@ -370,6 +371,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "commentReference")    
     public static class CommentReference
         implements Child
     {
@@ -450,6 +452,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "continuationSeparator")
     public static class ContinuationSeparator
         implements Child
     {
@@ -504,6 +507,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "cr")    
     public static class Cr
         implements Child
     {
@@ -558,6 +562,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "dayLong")    
     public static class DayLong
         implements Child
     {
@@ -612,6 +617,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "dayShort")    
     public static class DayShort
         implements Child
     {
@@ -666,6 +672,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "endnoteRef")    
     public static class EndnoteRef
         implements Child
     {
@@ -720,6 +727,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "footnoteRef")    
     public static class FootnoteRef
         implements Child
     {
@@ -774,6 +782,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "lastRenderedPageBreak")    
     public static class LastRenderedPageBreak
         implements Child
     {
@@ -828,6 +837,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "monthLong")    
     public static class MonthLong
         implements Child
     {
@@ -882,6 +892,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "monthShort")        
     public static class MonthShort
         implements Child
     {
@@ -936,6 +947,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "noBreakHyphen")        
     public static class NoBreakHyphen
         implements Child
     {
@@ -990,6 +1002,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "pgNum")            
     public static class PgNum
         implements Child
     {
@@ -1047,6 +1060,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "ptab")            
     public static class Ptab
         implements Child
     {
@@ -1179,6 +1193,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "separator")            
     public static class Separator
         implements Child
     {
@@ -1233,6 +1248,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "softHyphen")            
     public static class SoftHyphen
         implements Child
     {
@@ -1289,6 +1305,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "sym")            
     public static class Sym
         implements Child
     {
@@ -1395,6 +1412,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "tab")            
     public static class Tab
         implements Child
     {
@@ -1449,6 +1467,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "yearLong")            
     public static class YearLong
         implements Child
     {
@@ -1503,6 +1522,7 @@
      */
     @XmlAccessorType(XmlAccessType.FIELD)
     @XmlType(name = "")
+    @XmlRootElement(name = "yearShort")            
     public static class YearShort
         implements Child
     {
diff -u -r ./org/docx4j/wml/RPr.java ../patched/org/docx4j/wml/RPr.java
--- ./org/docx4j/wml/RPr.java	2009-10-10 09:56:02.000000000 +1100
+++ ../patched/org/docx4j/wml/RPr.java	2008-05-01 18:38:43.000000000 +1000
@@ -18,12 +18,12 @@
 
  */
 
-
 package org.docx4j.wml;
 
 import javax.xml.bind.Unmarshaller;
 import javax.xml.bind.annotation.XmlAccessType;
 import javax.xml.bind.annotation.XmlAccessorType;
+import javax.xml.bind.annotation.XmlRootElement;
 import javax.xml.bind.annotation.XmlTransient;
 import javax.xml.bind.annotation.XmlType;
 import org.jvnet.jaxb2_commons.ppp.Child;
@@ -91,6 +91,7 @@
     "oMath",
     "rPrChange"
 })
+@XmlRootElement(name = "rPr")
 public class RPr
     implements Child
 {
diff -u -r ./org/docx4j/wml/SdtBlock.java ../patched/org/docx4j/wml/SdtBlock.java
--- ./org/docx4j/wml/SdtBlock.java	2009-10-10 09:56:02.000000000 +1100
+++ ../patched/org/docx4j/wml/SdtBlock.java	2008-05-06 11:45:03.000000000 +1000
@@ -21,12 +21,15 @@
 
 package org.docx4j.wml;
 
+import javax.xml.bind.JAXBElement;
 import javax.xml.bind.Unmarshaller;
 import javax.xml.bind.annotation.XmlAccessType;
 import javax.xml.bind.annotation.XmlAccessorType;
 import javax.xml.bind.annotation.XmlRootElement;
 import javax.xml.bind.annotation.XmlTransient;
 import javax.xml.bind.annotation.XmlType;
+
+import org.apache.log4j.Logger;
 import org.jvnet.jaxb2_commons.ppp.Child;
 
 
@@ -63,6 +66,8 @@
 public class SdtBlock implements Child
 {
 
+	private static Logger log = Logger.getLogger(SdtBlock.class);		
+	
     protected SdtPr sdtPr;
     protected CTSdtEndPr sdtEndPr;
     protected SdtContentBlock sdtContent;
@@ -167,4 +172,103 @@
         setParent(parent);
     }
 
+    /* Replace any nested content controls with their content. */
+    public void flatten() {
+    	 
+		/*    	
+		<w:sdt xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main">
+			<w:sdtPr><w:tag w:val="30" /><w:id w:val="871785936" /></w:sdtPr>
+			<w:sdtContent>
+				<w:p w:rsidR="00283267" w:rsidRDefault="00E8712C">
+					<w:r><w:t>S</w:t></w:r>
+					<w:sdt><w:sdtPr><w:tag w:val="0" /><w:id w:val="589321610" /></w:sdtPr><w:sdtContent><w:r><w:t>Para1</w:t></w:r></w:sdtContent></w:sdt>
+				</w:p>
+			</w:sdtContent>
+		</w:sdt>
+		*/
+    	
+		log.info("Flattening sdt: " + sdtPr.getId().toString() );
+		boolean startAgain;
+    	do {
+        	startAgain = false;
+//        	java.util.Iterator it = sdtContent.getBlockLevelElements().iterator();
+//	    	while ( it.hasNext() ) {
+//	    		
+//	    		Object o = it.next();
+	    		
+	    	for (Object o : sdtContent.getEGContentBlockContent() ) {
+	    		
+	    		if (o instanceof SdtBlock) { // A block level SDT - but this doesn't happen
+	    			log.debug("Interesting .. detected BLOCK level nested sdt: " + ((SdtBlock)o).sdtPr.getId().toString() );
+	    			sdtContent.replaceElement(o, ((SdtBlock)o).getSdtContent().getEGContentBlockContent() );
+	    			// need to refresh the list we are iterating
+	    			startAgain = true;
+	    			break;
+	    		} else if ( o instanceof org.docx4j.wml.P ) {
+    				log.debug( "Paragraph object: ");
+    				org.docx4j.wml.P p = (org.docx4j.wml.P)o;   				
+    				flattenP(p);
+ 	    		} else if (o instanceof javax.xml.bind.JAXBElement) {
+	    			
+//	    			if ( ((JAXBElement)o).getDeclaredType().getName().equals("org.docx4j.wml.P") ) {
+//	    				log.debug( "Paragraph object: ");
+//	    				org.docx4j.wml.P p = (org.docx4j.wml.P)((JAXBElement)o).getValue();
+//	    				
+//	    				flattenP(p);
+//	    				// Is this necessary?
+//	    				((JAXBElement)o).setValue(p);
+//	    			} else {
+	    				
+	    				log.debug( "JAXB: " + ((JAXBElement)o).getValue().getClass().getName() );
+	    				
+//	    			}
+	    				
+	    		} else {
+	    			log.debug(o.getClass().getName() + ".. not an sdt");
+	    		}
+	    	}
+    	} while (startAgain);
+    }
+    	
+    public void flattenP(org.docx4j.wml.P p) {
+    	
+		/*    	
+				<w:p w:rsidR="00283267" w:rsidRDefault="00E8712C">
+					<w:r><w:t>S</w:t></w:r>
+					<w:sdt><w:sdtPr><w:tag w:val="0" /><w:id w:val="589321610" /></w:sdtPr><w:sdtContent><w:r><w:t>Para1</w:t></w:r></w:sdtContent></w:sdt>
+				</w:p>
+		*/
+    	
+		log.info("Flattening nested p " );
+		boolean startAgain;
+    	do {
+        	startAgain = false;
+	    	for (Object o : p.getParagraphContent() ) {
+	    		
+	    		if (o instanceof SdtRun) {  // This code path not used
+	    			log.debug(".. detected nested sdt " );
+	    			p.replaceElement(o, ((SdtRun)o).getSdtContent().getParagraphContent() );
+	    			// need to refresh the list we are iterating
+	    			startAgain = true;
+	    			break;
+	    		} else if (o instanceof javax.xml.bind.JAXBElement) {
+	    			
+	    			if ( ((JAXBElement)o).getDeclaredType().getName().equals("org.docx4j.wml.SdtRun") ) {
+	    				log.debug( ((JAXBElement)o).getDeclaredType().getName() + ".. detected SdtRun");
+	    				org.docx4j.wml.SdtRun sdtRun = (org.docx4j.wml.SdtRun)((JAXBElement)o).getValue();
+	    				p.replaceElement(o, sdtRun.getSdtContent().getParagraphContent() );
+	    			} else {
+	    				log.debug( ((JAXBElement)o).getDeclaredType().getName() + ".. not an sdt");	    				
+	    			}
+	    				
+	    		} else {
+	    			log.debug(o.getClass().getName() + ".. not an sdt");
+	    		}
+	    	}
+    	} while (startAgain);
+    	
+    	
+    }
+    
+    
 }
diff -u -r ./org/docx4j/wml/SdtContentBlock.java ../patched/org/docx4j/wml/SdtContentBlock.java
--- ./org/docx4j/wml/SdtContentBlock.java	2009-10-10 09:56:02.000000000 +1100
+++ ../patched/org/docx4j/wml/SdtContentBlock.java	2009-10-10 01:22:54.000000000 +1100
@@ -18,7 +18,6 @@
 
  */
 
-
 package org.docx4j.wml;
 
 import java.util.ArrayList;
@@ -29,10 +28,12 @@
 import javax.xml.bind.annotation.XmlAccessorType;
 import javax.xml.bind.annotation.XmlElementRef;
 import javax.xml.bind.annotation.XmlElementRefs;
+import javax.xml.bind.annotation.XmlRootElement;
 import javax.xml.bind.annotation.XmlTransient;
 import javax.xml.bind.annotation.XmlType;
 import org.docx4j.math.CTOMath;
 import org.docx4j.math.CTOMathPara;
+import org.apache.log4j.Logger;
 import org.jvnet.jaxb2_commons.ppp.Child;
 
 
@@ -179,4 +182,39 @@
         setParent(parent);
     }
 
+    public void replaceElement(Object current, List insertions) {
+
+    	int index = egContentBlockContent.indexOf(current);    	
+    	if (index > -1 ) {    		
+    		egContentBlockContent.addAll(index+1, insertions);  
+    		Object removed = egContentBlockContent.remove(index);
+    		// sanity check
+    		if (!current.equals(removed)) {
+    			log.error("removed wrong object?");
+    		}    		
+    	} else {
+    		// Not found
+    		log.error("Couldn't find replacement target.");
+    	}
+
+/*    	
+    	List<Object> newList = new ArrayList<Object>();
+    	
+    	for (Object o : getBlockLevelElements() ) {
+    		
+    		if (o.equals(current)) {
+    			log.debug(".. found target ");
+    			for (Object o2 : insertions ) {
+    				newList.add(o2);
+        			log.debug(".. .. inserted ");
+    			}
+    		} else {
+				newList.add(o);
+    		}
+    	}
+    	
+    	blockLevelElements = newList;
+    	*/ 
+    }
+    
 }
diff -u -r ./org/docx4j/wml/SdtPr.java ../patched/org/docx4j/wml/SdtPr.java
--- ./org/docx4j/wml/SdtPr.java	2009-10-10 09:56:02.000000000 +1100
+++ ../patched/org/docx4j/wml/SdtPr.java	2009-10-10 01:22:54.000000000 +1100
@@ -30,10 +30,12 @@
 import javax.xml.bind.annotation.XmlAttribute;
 import javax.xml.bind.annotation.XmlElementRef;
 import javax.xml.bind.annotation.XmlElementRefs;
+import javax.xml.bind.annotation.XmlRootElement;
 import javax.xml.bind.annotation.XmlTransient;
 import javax.xml.bind.annotation.XmlType;
 import org.jvnet.jaxb2_commons.ppp.Child;
 
+import org.apache.log4j.Logger;
 
 /**
  * <p>Java class for CT_SdtPr complex type.
@@ -209,6 +213,112 @@
         }
         return this.rPrOrAliasOrLock;
     }
+    
+    /**
+     * Gets the value of the id property.
+     * 
+     * @return
+     *     possible object is
+     *     {@link Id }
+     *     
+     */
+    public Id getId() {
+    	
+    	for (Object o : getRPrOrAliasOrLock()) {
+    		if ( o instanceof Id ) {
+    			log.debug("found id");
+    			return (Id)o;
+    		} 
+    	}
+    	
+        return null;
+    }
+
+    /**
+     * Sets the value of the id property.
+     * 
+     * @param value
+     *     allowed object is
+     *     {@link Id }
+     *     
+     */
+    public void setId(Id value) {
+    	
+    	Id existingId = getId(); 
+    	
+    	if (existingId!=null) {
+    		if (!existingId.equals(value)) {
+    			log.debug("Changing SDT ID from " + existingId + " to " + value);
+        		rPrOrAliasOrLock.remove(existingId);
+        		rPrOrAliasOrLock.add(value);
+    		}    	
+    		// else - they are the same, so do nothing
+    	} else {
+    		//ObjectFactory factory = new ObjectFactory();
+    		//JAXBElement idWrapper = factory.createSdtPrId(value);
+    		rPrOrAliasOrLock.add(value);
+    	}
+    }
+    
+    // Not generated!
+    public java.math.BigInteger setId() {
+
+    	Id id = new Id();
+    	java.math.BigInteger newIdVal = java.math.BigInteger.valueOf(Math.abs(new java.util.Random().nextInt()));
+    	id.setVal( newIdVal );
+    	setId(id);    	
+		log.warn("Generated random id: " + newIdVal.toString() );
+		return newIdVal;
+    }
+
+    /**
+     * Gets the value of the tag property.
+     * 
+     * @return
+     *     possible object is
+     *     {@link SdtPr.Tag }
+     *     
+     */
+    public Tag getTag() {
+    	
+    	for (Object o : getRPrOrAliasOrLock()) {
+    		if ( o instanceof Tag ) {
+    			log.debug("found tag");
+    			return (Tag)o;
+    		} 
+    	}
+    	
+        return null;
+    }
+
+    /**
+     * Sets the value of the tag property.
+     * 
+     * @param value
+     *     allowed object is
+     *     {@link SdtPr.Tag }
+     *     
+     */
+    public void setTag(Tag value) {
+        
+    	Tag existingTag = getTag(); 
+    	
+    	if (existingTag!=null) {
+    		if (!existingTag.equals(value)) {
+    			log.debug("Changing SDT ID from " + existingTag + " to " + value);
+        		rPrOrAliasOrLock.remove(existingTag);
+        		rPrOrAliasOrLock.add(value);
+    		}
+    		// else - they are the same, so do nothing
+    	} else {
+    		//ObjectFactory factory = new ObjectFactory();
+    		//JAXBElement idWrapper = factory.createSdtPrId(value);
+    		rPrOrAliasOrLock.add(value);
+    	}
+        
+        
+    }
+    
 
     /**
      * Gets the parent object in the object tree representing the unmarshalled xml document.
diff -u -r ./org/docx4j/wml/SectPr.java ../patched/org/docx4j/wml/SectPr.java
--- ./org/docx4j/wml/SectPr.java	2009-10-10 09:56:02.000000000 +1100
+++ ../patched/org/docx4j/wml/SectPr.java	2009-09-02 14:23:32.000000000 +1000
@@ -24,12 +24,14 @@
 import java.math.BigInteger;
 import java.util.ArrayList;
 import java.util.List;
+import javax.xml.bind.JAXBElement;
 import javax.xml.bind.Unmarshaller;
 import javax.xml.bind.annotation.XmlAccessType;
 import javax.xml.bind.annotation.XmlAccessorType;
 import javax.xml.bind.annotation.XmlAttribute;
 import javax.xml.bind.annotation.XmlElement;
 import javax.xml.bind.annotation.XmlElements;
+import javax.xml.bind.annotation.XmlRootElement;
 import javax.xml.bind.annotation.XmlTransient;
 import javax.xml.bind.annotation.XmlType;
 import org.jvnet.jaxb2_commons.ppp.Child;
@@ -81,6 +83,7 @@
     "printerSettings",
     "sectPrChange"
 })
+@XmlRootElement(name = "sectPr")
 public class SectPr
     implements Child
 {
diff -u -r ./org/docx4j/wml/Styles.java ../patched/org/docx4j/wml/Styles.java
--- ./org/docx4j/wml/Styles.java	2009-10-10 09:56:02.000000000 +1100
+++ ../patched/org/docx4j/wml/Styles.java	2009-09-02 14:23:32.000000000 +1000
@@ -667,7 +667,12 @@
              *     
              */
             public Boolean isQFormat() {
-                return qFormat;
+            	
+            	if (qFormat==null) {
+            		return ((Styles.LatentStyles)this.parent).isDefQFormat();
+            	} else {            	
+            		return qFormat;
+            	}
             }
 
             /**
diff -u -r ./org/docx4j/wml/Tabs.java ../patched/org/docx4j/wml/Tabs.java
--- ./org/docx4j/wml/Tabs.java	2009-10-10 09:56:02.000000000 +1100
+++ ../patched/org/docx4j/wml/Tabs.java	2009-06-22 14:41:30.000000000 +1000
@@ -27,6 +27,7 @@
 import javax.xml.bind.annotation.XmlAccessType;
 import javax.xml.bind.annotation.XmlAccessorType;
 import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlRootElement;
 import javax.xml.bind.annotation.XmlTransient;
 import javax.xml.bind.annotation.XmlType;
 import org.jvnet.jaxb2_commons.ppp.Child;
@@ -55,6 +56,7 @@
 @XmlType(name = "CT_Tabs", propOrder = {
     "tab"
 })
+@XmlRootElement(name = "tabs")
 public class Tabs
     implements Child
 {
diff -u -r ./org/docx4j/wml/Tbl.java ../patched/org/docx4j/wml/Tbl.java
--- ./org/docx4j/wml/Tbl.java	2009-10-10 09:56:02.000000000 +1100
+++ ../patched/org/docx4j/wml/Tbl.java	2009-10-10 01:22:54.000000000 +1100
@@ -18,7 +18,6 @@
 
  */
 
-
 package org.docx4j.wml;
 
 import java.util.ArrayList;
@@ -30,6 +29,7 @@
 import javax.xml.bind.annotation.XmlElement;
 import javax.xml.bind.annotation.XmlElementRef;
 import javax.xml.bind.annotation.XmlElementRefs;
+import javax.xml.bind.annotation.XmlRootElement;
 import javax.xml.bind.annotation.XmlTransient;
 import javax.xml.bind.annotation.XmlType;
 import org.docx4j.math.CTOMath;
@@ -64,6 +64,7 @@
     "tblGrid",
     "egContentRowContent"
 })
+@XmlRootElement(name = "tbl")
 public class Tbl
     implements Child
 {
diff -u -r ./org/docx4j/wml/Tc.java ../patched/org/docx4j/wml/Tc.java
--- ./org/docx4j/wml/Tc.java	2009-10-10 09:56:02.000000000 +1100
+++ ../patched/org/docx4j/wml/Tc.java	2009-10-10 01:22:54.000000000 +1100
@@ -29,6 +29,7 @@
 import javax.xml.bind.annotation.XmlAccessorType;
 import javax.xml.bind.annotation.XmlElementRef;
 import javax.xml.bind.annotation.XmlElementRefs;
+import javax.xml.bind.annotation.XmlRootElement;
 import javax.xml.bind.annotation.XmlTransient;
 import javax.xml.bind.annotation.XmlType;
 import org.docx4j.math.CTOMath;
diff -u -r ./org/docx4j/wml/Text.java ../patched/org/docx4j/wml/Text.java
--- ./org/docx4j/wml/Text.java	2009-10-10 09:56:02.000000000 +1100
+++ ../patched/org/docx4j/wml/Text.java	2008-04-09 13:11:16.000000000 +1000
@@ -18,12 +18,12 @@
 
  */
 
-
 package org.docx4j.wml;
 
 import javax.xml.bind.Unmarshaller;
 import javax.xml.bind.annotation.XmlAccessType;
 import javax.xml.bind.annotation.XmlAccessorType;
+import javax.xml.bind.annotation.XmlRootElement;
 import javax.xml.bind.annotation.XmlAttribute;
 import javax.xml.bind.annotation.XmlTransient;
 import javax.xml.bind.annotation.XmlType;
@@ -54,6 +54,7 @@
 @XmlType(name = "CT_Text", propOrder = {
     "value"
 })
+@XmlRootElement(name = "t")
 public class Text
     implements Child
 {
